db-service/test/cqn2sql/__snapshots__/upsert.test.js.snap

```diff
@@ -29,6 +29,6 @@ exports[`upsert test with rows (quoted) 1`] = `
       "[[1,null,2]]",
     ],
   ],
-  "sql": "INSERT INTO """Foo2Quoted""" ("""ID""","""name""","""a""") SELECT value->>'$[0]',value->>'$[1]',value->>'$[2]' FROM json_each(?) WHERE true ON CONFLICT("""ID""") DO UPDATE SET """name""" = excluded."""name""","""a""" = excluded."""a"""",
+  "sql": "INSERT INTO """Foo2Quoted""" ("""ID""","""name""","""a""") SELECT (CASE WHEN OLD."""ID""" IS NULL THEN value->>'$."""ID"""' ELSE (CASE WHEN json_type(value,'$."""ID"""') IS NULL THEN OLD."""ID""" ELSE value->>'$."""ID"""' END) END) as """ID""",(CASE WHEN OLD."""ID""" IS NULL THEN value->>'$."""name"""' ELSE (CASE WHEN json_type(value,'$."""name"""') IS NULL THEN OLD."""name""" ELSE value->>'$."""name"""' END) END) as """name""",(CASE WHEN OLD."""ID""" IS NULL THEN value->>'$."""a"""' ELSE (CASE WHEN json_type(value,'$."""a"""') IS NULL THEN OLD."""a""" ELSE value->>'$."""a"""' END) END) as """a""" FROM (SELECT value, value->>'$."""ID"""' as """ID""" from json_each(?)) as NEW LEFT JOIN """Foo2Quoted""" AS OLD ON NEW."""ID"""=OLD."""ID""" WHERE TRUE ON CONFLICT("ID") DO UPDATE SET """name""" = excluded."""name""","""a""" = excluded."""a"""",
```

Comment:
For keys, I'm a bit confused by

```sql
(CASE 
  WHEN OLD."ID" IS NULL THEN value->>'$."ID"' 
  ELSE 
  (CASE 
    WHEN json_type(value, '$."ID"') IS NULL THEN OLD."ID" 
    ELSE value->>'$."ID"' 
    END) 
  END) AS "ID"
```

We demand that all keys are provided, otherwise `UPSERT` doesn't make sense. Hence, for key fields, we can always take the value from the JSON input, no?

```sql
value->>'$."ID"'
```